### ISSUE TITLE
PHPUnit Coverage Reference

### ISSUE THREAD

**Issue by AndrolGenhald** [vimeo/psalm#7327]
When working with Psalm's code, it's sometimes difficult to tell what a specific piece of code is for. If we had a simple way to mark each line of code with the tests that cover it, it would be a great reference tool for what every part of the code is doing. Currently if I really want to see why some piece of code exists and I can't easily figure it out, I'll often comment it out and run the tests, but this takes 2 minutes with paratest on my 12 core/24 thread machine, and I imagine it could take a lot longer for others.
Maybe this already exists and I'm just not aware of it.

I also just found out about [filtering data providers](https://phpunit.de/manual/6.5/en/textui.html#textui.examples.filter-shortcuts), I wasn't aware that was possible. Once a test that covers a piece of code is located it's easy enough to run only that test with xdebug to step through and see exactly what happens, if it's not already obvious from the test itself.

**Comment by orklah** [comment]
I'm pretty sure we had that at one point but it was disabled due to the time it took to process it.

**Comment by AndrolGenhald** [comment]
Do I need to go looking through PHPUnit documentation? Even if I have to run it locally and it takes half an hour that's fine, just running it once a week on master would already be incredibly helpful. Most of the code doesn't change that often, so even an out of date reference will be correct almost all the time.

**Comment by orklah** [comment]
I'm not actually sure how that looked, it was disabled before I started contributing I believe and I don't know much about phpunit on the coverage topic. Maybe @weirdan could tell us more if he remembers, otherwise yeah, you could take a look, it could be an interesting tool

**Comment by weirdan** [comment]
Related: #6800

**Comment by AndrolGenhald** [comment]
@weirdan Thanks! I probably should have searched a bit more. I had to install `php-pcov` and set the environment variable `XDEBUG_MODE=coverage`, but once I figured that out it ran in about 18 minutes and produced HTML coverage files totaling 7.6GiB.

`XDEBUG_MODE=coverage time vendor/bin/paratest --coverage-html build/phpunit/coverage-html`

This will be useful to have locally, even if it's too big to upload somewhere.
